File: snippets/security/authentication/session_management.ts
/**
 * Secure Session Management (TypeScript)
 * Production-ready session handling with Redis backend
 */
import { createClient, RedisClientType } from 'redis';
import { randomBytes, createHash, createHmac } from 'crypto';

interface SessionData {
  user_id: number;
  created_at: string;
  last_activity: string;
  data: Record<string, any>;
}

interface CookieOptions {
  maxAge: number;
  path: string;
  secure: boolean;
  httpOnly: boolean;
  sameSite: 'Strict' | 'Lax' | 'None';
  domain?: string;
}

export class SessionManager {
  private redis: RedisClientType;
  private sessionLifetime: number;
  private sessionPrefix: string;

  /**
   * Initialize session manager
   * @param redis - Redis client instance
   * @param sessionLifetime - Session lifetime in seconds
   * @param sessionPrefix - Prefix for Redis keys
   */
  constructor(
    redis: RedisClientType,
    sessionLifetime: number = 3600,
    sessionPrefix: string = 'session:'
  ) {
    this.redis = redis;
    this.sessionLifetime = sessionLifetime;
    this.sessionPrefix = sessionPrefix;
  }

  /**
   * Generate cryptographically secure session ID
   * @returns Session ID string
   */
  generateSessionId(): string {
    const randomData = randomBytes(32);
    const sessionId = createHash('sha256').update(randomData).digest('hex');
    return sessionId;
  }

  /**
   * Create a new session
   * @param userId - User ID
   * @param data - Additional session data
   * @returns Session ID
   */
  async createSession(
    userId: number,
    data?: Record<string, any>
  ): Promise<string> {
    const sessionId = this.generateSessionId();
    const now = new Date().toISOString();

    const sessionData: SessionData = {
      user_id: userId,
      created_at: now,
      last_activity: now,
      data: data || {},
    };

    const key = `${this.sessionPrefix}${sessionId}`;

    await this.redis.setEx(
      key,
      this.sessionLifetime,
      JSON.stringify(sessionData)
    );

    return sessionId;
  }

  /**
   * Retrieve session data
   * @param sessionId - Session ID
   * @returns Session data or null if not found
   */
  async getSession(sessionId: string): Promise<SessionData | null> {
    const key = `${this.sessionPrefix}${sessionId}`;
    const data = await this.redis.get(key);

    if (!data) {
      return null;
    }

    const sessionData: SessionData = JSON.parse(data);

    // Update last activity
    sessionData.last_activity = new Date().toISOString();

    await this.redis.setEx(
      key,
      this.sessionLifetime,
      JSON.stringify(sessionData)
    );

    return sessionData;
  }

  /**
   * Update session data
   * @param sessionId - Session ID
   * @param data - Data to update
   * @returns True if successful
   */
  async updateSession(
    sessionId: string,
    data: Record<string, any>
  ): Promise<boolean> {
    const sessionData = await this.getSession(sessionId);

    if (!sessionData) {
      return false;
    }

    sessionData.data = { ...sessionData.data, ...data };
    sessionData.last_activity = new Date().toISOString();

    const key = `${this.sessionPrefix}${sessionId}`;

    await this.redis.setEx(
      key,
      this.sessionLifetime,
      JSON.stringify(sessionData)
    );

    return true;
  }

  /**
   * Delete a session (logout)
   * @param sessionId - Session ID
   * @returns True if deleted
   */
  async deleteSession(sessionId: string): Promise<boolean> {
    const key = `${this.sessionPrefix}${sessionId}`;
    const result = await this.redis.del(key);
    return result > 0;
  }

  /**
   * Delete all sessions for a user
   * @param userId - User ID
   * @returns Number of sessions deleted
   */
  async deleteUserSessions(userId: number): Promise<number> {
    const pattern = `${this.sessionPrefix}*`;
    let deleted = 0;

    for await (const key of this.redis.scanIterator({ MATCH: pattern })) {
      const data = await this.redis.get(key);

      if (data) {
        const sessionData: SessionData = JSON.parse(data);

        if (sessionData.user_id === userId) {
          await this.redis.del(key);
          deleted++;
        }
      }
    }

    return deleted;
  }

  /**
   * Check if session is valid
   * @param sessionId - Session ID
   * @returns True if session exists and is valid
   */
  async validateSession(sessionId: string): Promise<boolean> {
    const session = await this.getSession(sessionId);
    return session !== null;
  }

  /**
   * Count active sessions for a user
   * @param userId - User ID
   * @returns Number of active sessions
   */
  async getSessionCount(userId: number): Promise<number> {
    const pattern = `${this.sessionPrefix}*`;
    let count = 0;

    for await (const key of this.redis.scanIterator({ MATCH: pattern })) {
      const data = await this.redis.get(key);

      if (data) {
        const sessionData: SessionData = JSON.parse(data);

        if (sessionData.user_id === userId) {
          count++;
        }
      }
    }

    return count;
  }
}

export class SecureCookieSession {
  /**
   * Get secure cookie options
   * @param maxAge - Cookie lifetime in seconds
   * @param domain - Cookie domain
   * @param path - Cookie path
   * @returns Cookie options
   */
  static getSecureCookieOptions(
    maxAge: number = 3600,
    domain?: string,
    path: string = '/'
  ): CookieOptions {
    const options: CookieOptions = {
      maxAge,
      path,
      secure: true, // HTTPS only
      httpOnly: true, // Not accessible via JavaScript
      sameSite: 'Lax', // CSRF protection
    };

    if (domain) {
      options.domain = domain;
    }

    return options;
  }

  /**
   * Sign cookie value for integrity
   * @param value - Cookie value
   * @param secret - Secret key
   * @returns Signed value
   */
  static signCookieValue(value: string, secret: string): string {
    const signature = createHmac('sha256', secret)
      .update(value)
      .digest('hex');

    return `${value}.${signature}`;
  }

  /**
   * Verify and extract cookie value
   * @param signedValue - Signed cookie value
   * @param secret - Secret key
   * @returns Original value if valid, null otherwise
   */
  static verifyCookieSignature(
    signedValue: string,
    secret: string
  ): string | null {
    try {
      const lastDotIndex = signedValue.lastIndexOf('.');
      if (lastDotIndex === -1) {
        return null;
      }

      const value = signedValue.substring(0, lastDotIndex);
      const signature = signedValue.substring(lastDotIndex + 1);

      const expectedSignature = createHmac('sha256', secret)
        .update(value)
        .digest('hex');

      // Constant-time comparison
      if (signature === expectedSignature) {
        return value;
      }

      return null;
    } catch {
      return null;
    }
  }

  /**
   * Format cookie header string
   * @param name - Cookie name
   * @param value - Cookie value
   * @param options - Cookie options
   * @returns Set-Cookie header value
   */
  static formatCookieHeader(
    name: string,
    value: string,
    options: CookieOptions
  ): string {
    const parts = [`${name}=${value}`];

    if (options.maxAge) {
      parts.push(`Max-Age=${options.maxAge}`);
    }

    if (options.path) {
      parts.push(`Path=${options.path}`);
    }

    if (options.domain) {
      parts.push(`Domain=${options.domain}`);
    }

    if (options.secure) {
      parts.push('Secure');
    }

    if (options.httpOnly) {
      parts.push('HttpOnly');
    }

    if (options.sameSite) {
      parts.push(`SameSite=${options.sameSite}`);
    }

    return parts.join('; ');
  }
}

// Example usage
if (require.main === module) {
  (async () => {
    // Initialize Redis client
    const redis = createClient({
      url: 'redis://localhost:6379',
    });

    await redis.connect();

    // Create session manager
    const sessionMgr = new SessionManager(redis);

    // Create a new session
    const userId = 123;
    const sessionId = await sessionMgr.createSession(userId, {
      role: 'admin',
      ip: '192.168.1.1',
    });

    console.log(`Created session: ${sessionId}`);

    // Retrieve session
    const session = await sessionMgr.getSession(sessionId);
    console.log('Session data:', session);

    // Update session
    await sessionMgr.updateSession(sessionId, { last_page: '/dashboard' });

    // Validate session
    const isValid = await sessionMgr.validateSession(sessionId);
    console.log(`Session valid: ${isValid}`);

    // Get secure cookie options
    const cookieOptions = SecureCookieSession.getSecureCookieOptions();
    console.log('Cookie options:', cookieOptions);

    // Sign cookie
    const signedCookie = SecureCookieSession.signCookieValue(
      sessionId,
      'secret_key'
    );
    console.log(`Signed cookie: ${signedCookie.substring(0, 50)}...`);

    // Delete session
    await sessionMgr.deleteSession(sessionId);
    console.log('Session deleted');

    await redis.disconnect();
  })();
}

File: snippets/security/authentication/session_management.py
"""
Secure Session Management
Production-ready session handling with Redis backend
"""
import secrets
import json
import hashlib
import hmac
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
import redis


class SessionManager:
    """Secure session management with Redis"""

    def __init__(
        self,
        redis_client: redis.Redis,
        session_lifetime: int = 3600,  # 1 hour
        session_prefix: str = "session:"
    ):
        """
        Initialize session manager

        Args:
            redis_client: Redis client instance
            session_lifetime: Session lifetime in seconds
            session_prefix: Prefix for Redis keys
        """
        self.redis = redis_client
        self.session_lifetime = session_lifetime
        self.session_prefix = session_prefix

    def generate_session_id(self) -> str:
        """
        Generate cryptographically secure session ID

        Returns:
            Session ID string
        """
        # Generate 32 bytes of random data
        random_bytes = secrets.token_bytes(32)

        # Hash with SHA256 for consistency
        session_id = hashlib.sha256(random_bytes).hexdigest()

        return session_id

    def create_session(self, user_id: int, data: Optional[Dict] = None) -> str:
        """
        Create a new session

        Args:
            user_id: User ID
            data: Additional session data

        Returns:
            Session ID
        """
        session_id = self.generate_session_id()
        session_data = {
            "user_id": user_id,
            "created_at": datetime.utcnow().isoformat(),
            "last_activity": datetime.utcnow().isoformat(),
            "data": data or {}
        }

        # Store in Redis with expiration
        key = f"{self.session_prefix}{session_id}"
        self.redis.setex(
            key,
            self.session_lifetime,
            json.dumps(session_data)
        )

        return session_id

    def get_session(self, session_id: str) -> Optional[Dict[str, Any]]:
        """
        Retrieve session data

        Args:
            session_id: Session ID

        Returns:
            Session data or None if not found
        """
        key = f"{self.session_prefix}{session_id}"
        data = self.redis.get(key)

        if data:
            session_data = json.loads(data)

            # Update last activity
            session_data["last_activity"] = datetime.utcnow().isoformat()
            self.redis.setex(
                key,
                self.session_lifetime,
                json.dumps(session_data)
            )

            return session_data

        return None

    def update_session(
        self,
        session_id: str,
        data: Dict[str, Any]
    ) -> bool:
        """
        Update session data

        Args:
            session_id: Session ID
            data: Data to update

        Returns:
            True if successful
        """
        session_data = self.get_session(session_id)

        if session_data:
            session_data["data"].update(data)
            session_data["last_activity"] = datetime.utcnow().isoformat()

            key = f"{self.session_prefix}{session_id}"
            self.redis.setex(
                key,
                self.session_lifetime,
                json.dumps(session_data)
            )
            return True

        return False

    def delete_session(self, session_id: str) -> bool:
        """
        Delete a session (logout)

        Args:
            session_id: Session ID

        Returns:
            True if deleted
        """
        key = f"{self.session_prefix}{session_id}"
        return bool(self.redis.delete(key))

    def delete_user_sessions(self, user_id: int) -> int:
        """
        Delete all sessions for a user

        Args:
            user_id: User ID

        Returns:
            Number of sessions deleted
        """
        pattern = f"{self.session_prefix}*"
        deleted = 0

        for key in self.redis.scan_iter(match=pattern):
            data = self.redis.get(key)
            if data:
                session_data = json.loads(data)
                if session_data.get("user_id") == user_id:
                    self.redis.delete(key)
                    deleted += 1

        return deleted

    def validate_session(self, session_id: str) -> bool:
        """
        Check if session is valid

        Args:
            session_id: Session ID

        Returns:
            True if session exists and is valid
        """
        return self.get_session(session_id) is not None

    def get_session_count(self, user_id: int) -> int:
        """
        Count active sessions for a user

        Args:
            user_id: User ID

        Returns:
            Number of active sessions
        """
        pattern = f"{self.session_prefix}*"
        count = 0

        for key in self.redis.scan_iter(match=pattern):
            data = self.redis.get(key)
            if data:
                session_data = json.loads(data)
                if session_data.get("user_id") == user_id:
                    count += 1

        return count


class SecureCookieSession:
    """Secure cookie-based session helper"""

    @staticmethod
    def get_secure_cookie_options(
        max_age: int = 3600,
        domain: Optional[str] = None,
        path: str = "/"
    ) -> Dict[str, Any]:
        """
        Get secure cookie options

        Args:
            max_age: Cookie lifetime in seconds
            domain: Cookie domain
            path: Cookie path

        Returns:
            Cookie options dictionary
        """
        options = {
            "max_age": max_age,
            "path": path,
            "secure": True,  # HTTPS only
            "httponly": True,  # Not accessible via JavaScript
            "samesite": "Lax"  # CSRF protection
        }

        if domain:
            options["domain"] = domain

        return options

    @staticmethod
    def sign_cookie_value(value: str, secret: str) -> str:
        """
        Sign cookie value for integrity

        Args:
            value: Cookie value
            secret: Secret key

        Returns:
            Signed value
        """
        signature = hmac.new(
            secret.encode(),
            value.encode(),
            hashlib.sha256
        ).hexdigest()

        return f"{value}.{signature}"

    @staticmethod
    def verify_cookie_signature(
        signed_value: str,
        secret: str
    ) -> Optional[str]:
        """
        Verify and extract cookie value

        Args:
            signed_value: Signed cookie value
            secret: Secret key

        Returns:
            Original value if valid, None otherwise
        """
        try:
            value, signature = signed_value.rsplit(".", 1)

            expected_signature = hmac.new(
                secret.encode(),
                value.encode(),
                hashlib.sha256
            ).hexdigest()

            if secrets.compare_digest(signature, expected_signature):
                return value

        except ValueError:
            pass

        return None


# Example usage
if __name__ == "__main__":
    # Initialize Redis client
    redis_client = redis.Redis(
        host='localhost',
        port=6379,
        db=0,
        decode_responses=False
    )

    # Create session manager
    session_mgr = SessionManager(redis_client)

    # Create a new session
    user_id = 123
    session_id = session_mgr.create_session(
        user_id,
        data={"role": "admin", "ip": "192.168.1.1"}
    )
    print(f"Created session: {session_id}")

    # Retrieve session
    session = session_mgr.get_session(session_id)
    print(f"Session data: {session}")

    # Update session
    session_mgr.update_session(session_id, {"last_page": "/dashboard"})

    # Validate session
    is_valid = session_mgr.validate_session(session_id)
    print(f"Session valid: {is_valid}")

    # Get secure cookie options
    cookie_options = SecureCookieSession.get_secure_cookie_options()
    print(f"Cookie options: {cookie_options}")

    # Delete session
    session_mgr.delete_session(session_id)
    print(f"Session deleted")
